arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collector.kt

```diff
@@ -0,0 +1,237 @@
+package arrow.collectors
+
+/**
+ * Defines how the [collect] function may run the collector.
+ */
+public enum class Characteristics {
+  /**
+   * The flow may be collected using several workers.
+   */
+  CONCURRENT,
+
+  /**
+   * The final step of the collector simply returns the accumulator.
+   * That means that final call may be replaced by a cast.
+   */
+  IDENTITY_FINISH,
+
+  /**
+   * The order in which elements are accumulated doesn't matter for the end result.
+   */
+  UNORDERED;
+
+  public companion object {
+    public val IDENTITY: Set<Characteristics> = setOf(IDENTITY_FINISH)
+    public val IDENTITY_UNORDERED: Set<Characteristics> = setOf(IDENTITY_FINISH, UNORDERED)
+    public val CONCURRENT_UNORDERED: Set<Characteristics> = setOf(CONCURRENT, UNORDERED)
+    public val IDENTITY_CONCURRENT: Set<Characteristics> = setOf(IDENTITY_FINISH, CONCURRENT)
+    public val IDENTITY_CONCURRENT_UNORDERED: Set<Characteristics> = setOf(IDENTITY_FINISH, CONCURRENT, UNORDERED)
+  }
+}
+
+public typealias Collector<Value, Result> = CollectorI<*, Value, Result>
+public typealias NonSuspendCollector<Value, Result> = NonSuspendCollectorI<*, Value, Result>
+
+/**
+ * A [Collector] accumulates information from elements
+ * coming from some data source, usually a [kotlinx.coroutines.flow.Flow] or [Iterable].
+ *
+ * The accumulation is done in three phases:
+ * - Initialization of some (mutable) accumulator ([supply])
+ * - Updating the accumulator with each value ([accumulate])
+ * - Finalize the work, and extract the final result ([finish])
+ *
+ * This interface is heavily influenced by
+ * Java's [`Collector`](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collector.html)
+ * and Haskell's [`foldl` library](https://hackage.haskell.org/package/foldl/docs/Control-Foldl.html).
+ */
+public interface CollectorI<InternalAccumulator, in Value, out Result> {
```

Comment:
**Nit:** Not a huge fan of `CollectorI` but not sure if something else makes sense.. I guess `InternalAccumulator` is `State`? What about `StateCollector<State, in Value, out Result>`, or `StatefulCollector`? Not sure if this makes sense though, since all `Collector` have `State`. We had a similar issue with Schedule when we originally designed it, but don't think we can take the same/similar approach to solve this issue.

This is just a _nit_, feel free to ignore. For me it's also fine without any changes, just wanted to leave some thoughts.